Add a lot more information to SB fatal errors

saethlin opened the pull request:
In fatal errors, this clarifies the difference between a tag not being present in the borrow stack at all, and the tag being present but granting SRO. It also introduces a little notation for memory ranges so we can mention to the user that the span may point to code that operates on multiple memory locations, but we are reporting an error at a particular offset.

This also gets rid of the unqualified phrase "the borrow stack" in errors, and clarifies that it is the borrow stack _for some location_.

The crate `pdqselect` v0.1.1:
Before:
```
2103 |     unsafe { copy_nonoverlapping(src, dst, count) }
     |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no item granting read access to tag <2357> at alloc1029 found in borrow stack.
```
After:
```
2103 |     unsafe { copy_nonoverlapping(src, dst, count) }
     |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |              |
     |              attempting a read access using <2357> at alloc1029[0x0], but that tag does not exist in the borrow stack for this location
     |              this error occurs as part of an access at alloc1029[0x0..0x4]
```

And the crate `half` v1.8.2
Before:
```
131 |     unsafe { &mut *ptr::slice_from_raw_parts_mut(data, len) }
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ trying to reborrow for Unique at alloc1051, but parent tag <2091> does not have an appropriate item in the borrow stack
```
After:
```
131 |     unsafe { &mut *ptr::slice_from_raw_parts_mut(data, len) }
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |              |
    |              trying to reborrow <2091> for Unique permission at alloc1051[0x0], but that tag only grants SharedReadOnly permission for this location
    |              this error occurs as part of a reborrow at alloc1051[0x0..0x6]
```

saethlin commented:
Also, this is going to break the tests. I could do with some advice on fixing them. I'm not entirely sure what the methodology is.

RalfJung commented:
Thanks a lot for the PR!

Just as a heads-up, I am busy with job interviews for probably at least another month, so it will be a while until I will be able to review this PR.

saethlin commented:
Hm, I ran into this with `ascii` version 1.0.0. Not yet sure what's going on here, but this looks bug-like
```
test ascii_str::tests::as_mut_ascii_str ... note: tracking was triggered
    --> src/ascii_str.rs:998:16
     |
998  |             Ok(slice.into())
     |                ^^^^^^^^^^^^ created tag <271733>, granting Unique over alloc100517[0x0..0x2]
     |

... no other diagnostics...

error: Undefined Behavior: trying to reborrow for SharedReadOnly over alloc100517[0x0..0x2], but parent tag <271733> does not grant the required permission over this memory range
    --> /home/ben/.rustup/toolchains/miri/lib/rustlib/src/rust/library/core/src/cmp.rs:1538:27
     |
1538 |             PartialEq::eq(*self, *other)
     |                           ^^^^^ trying to reborrow for SharedReadOnly over alloc100517[0x0..0x2], but parent tag <271733> does not grant the required permission over this memory range
     |
```

saethlin commented:
You of course know SB better than I do, but this looks to me like a bug in Miri, probably a bug in this PR. I feel like a pointer that grants Unique over alloc100517[0x0..0x2] should be able to be reborrows into SRO over that same memory range.

I'm going to try minimizing whatever is causing this, just posting here for general awareness that this PR may be buggy.

RalfJung commented on src/diagnostics.rs:
```diff
@@ -58,7 +58,7 @@ impl MachineStopType for TerminationInfo {}
 /// Miri specific diagnostics
 pub enum NonHaltingDiagnostic {
-    CreatedPointerTag(NonZeroU64),
+    CreatedPointerTag(NonZeroU64, Permission, AllocId, AllocRange),
```
Now that there are so many fields, it seems better to use a struct variant and give them names and comments explaining their purpose.

RalfJung commented on src/stacked_borrows.rs:
```diff
@@ -416,19 +417,54 @@ impl<'tcx> Stack {
-                new.perm, dbg_ptr, derived_from,
-            )))?;
+        let granting_idx = self.find_granting(access, derived_from).ok_or_else(|| {
+            let action = format!(
```
In order to separate concerns, is it possible to move all this code that exists solely for better diagnostics into a separate function?

RalfJung commented on src/stacked_borrows.rs:
```diff
@@ -565,8 +613,8 @@ impl Stacks {
-        self.for_each(range, move |offset, stack| {
-            stack.access(AccessKind::Read, tag, Pointer::new(alloc_id, offset), global)
+        self.for_each(range, move |_offset, stack| {
+            stack.access(AccessKind::Read, tag, (alloc_id, range), global)
```
This loses precision -- `range` will be the range of the *entire* access, but `access` here is called for just a particular sub-range starting at `offset`. So the resulting error might even be factually wrong, by stating a too large range.

saethlin replied:
The memory access that we're handling is possibly granted by `tag` over the entirety of `range` and of the same kind throughout. It doesn't matter where in that range the tag stops granting access, what I'm suggesting reporting is that `tag` doesn't grant the permission over all of `range`.

Perhaps it would be _more_ precise to state exactly what range the tag doesn't grant sufficient permission for, but I don't see how the diagnostics this reports are wrong.

RalfJung replied:
Fair -- I guess it depends on how one interprets "tag X is invalid for access at start..end", i.e. whether there is an implicit universal or existential quantification.

But then the question remains, why regress the currently more precise error messages? If you want to report a range, we can make `for_each` provide the precise range.

saethlin replied:
Can you come up with some code where this improved precision changes the diagnostic? I feel like this has something to do with freezing, but I can't come up with an example at the moment.

RalfJung replied:
No, freezing won't trigger this. You need to get the underlying `RangeMap` to split the ranges because the stacks diverge. Roughly like:
```rust
let mut pair = (0i32, 1i32);
let ptr = &mut pair as *mut _;
pair.0 += 1;
let _val = *ptr; // now just the range 0..4 of locations is "bad"
```

saethlin replied:
I'm still confused. I modified the program to this, to provoke an error:
```rust
fn main() {
    let mut pair = (0i32, 1i32);
    let ptr = &mut pair as *mut _;
    let another = &mut pair.1;
    unsafe {
        *ptr = (1i32, 1i32);
    }
}
```
But the diagnostic comes out as:
```
 --> src/main.rs:6:9
  |
6 |         *ptr = (1i32, 1i32);
  |         ^^^^^^^^^^^^^^^^^^^ attempting a write access of alloc953[0x4..0x8] via tag <untagged>, but <untagged> does not exist in the borrow stack for this memory range
  |
```
That seems correct to me.

If I modify the program a bit to be something I'm sure will write across a range which is split:
```rust
fn main() {
    let mut pair = [0i32, 1i32];
    let ptr = &mut pair as *mut i32 as *mut i64;
    let another = &mut pair[1];
    unsafe {
        *ptr = 0i64;
    }
}
```
Then I get this:
```
 --> src/main.rs:6:9
  |
6 |         *ptr = 0i64;
  |         ^^^^^^^^^^^ attempting a write access of alloc953[0x0..0x8] via tag <untagged>, but <untagged> does not exist in the borrow stack for this memory range
  |
```
It sounds to me like you want that to say `alloc953[0x4..0x8]`, but I think a user is likely to assume that's a Miri bug. This is clearly a write of size 8. Why is Miri telling me about a write of size 4?

saethlin replied:
I need to sleep, but questions like this make me want to draw diagrams of the allocation in question. Maybe that's a good thing to do eventually....

RalfJung replied:
> `*ptr = (1i32, 1i32)`

this gets compiled to two separate assignments:
```
((*_2).0: i32) = const 1_i32;    // scope 4 at [src/main.rs:6:9: 6:28]
((*_2).1: i32) = const 1_i32;    // scope 4 at [src/main.rs:6:9: 6:28]
```
That's why you are not seeing the range you are expecting.
Use something like this instead:
```rust
let newval = (1i32, 1i32);
*ptr = newval
```

> I modified the program to this, to provoke an error:

FWIW my example [also provokes an error](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e092a4527770a4175ebefce547263d82)?

> It sounds to me like you want that to say alloc953[0x4..0x8], but I think a user is likely to assume that's a Miri bug. This is clearly a write of size 8. Why is Miri telling me about a write of size 4?

Aliasing is strictly per-location, so an 8-byte access are simply 8 independent 1-byte accesses. It is for pure performance reasons that Miri groups neighboring borrow stacks together and often handles entire many-byte accesses via a single borrow stack operation.

saethlin replied:
My concern is printing a span which _clearly_ points to an access of size 8, then saying something about a smaller range. I think the best approach is to print the exact range in the error, and if the range was split, include a note which explains to the user that this error is just for the first part of the whole operation that Miri has detected is invalid. Does that sound good to you?

RalfJung replied:
That would be a possibility.

However I increasingly think we are going down the wrong path here. Stacked Borrows should be thought of as having an independent borrow stack for each memory location. The fact that Miri fuses neighboring stacks is an optimization and should not affect user-visible behavior. As such, a Stacked Borrows violation always occurs at a *single* memory address, never at a range -- and IMO that is how we should report it.

saethlin replied:
What about this:
```
  |
6 |         *ptr = 0;
  |         ^^^^^^^^ attempting a write access at alloc953+0x4 via tag <1234>, but <1234> does not exist in the borrow stack for this address

note: this error was detected while handling a write to alloc953[0x0..0x8]
```
The error is exactly correct, and in combination with the note, the user can reasonably deduce what is going on here without fishing through their source code to figure out the size of the pointed-to type.

RalfJung replied:
I assume you are concerned that putting the entire write range into the main message together with the exact address will make that message too long?

We don't have such notes in other errors, but also most errors don't contain nearly as much information as Stacked Borrows errors, so this is fine for me. It might be tricky to implement though.^^

The fact that it talks about a "write" twice seems somewhat redundant, not sure if that is needed.

saethlin replied:
Yeah, I'm concerned about the size of the errors. I have yet more ideas on things to add, and already they are big enough that adding to the one long line is becoming an issue.

I implemented two versions of this. One just sticks another line into the existing `helps` machinery, but that generates a whole extra span. The latest commit moves the extra line into the original span, with `span_label` and looks like this:
```
error: Undefined Behavior: trying to reborrow alloc953+0x0 with Unique permission, but parent tag <1981> only grants SharedReadOnly permission for this memory
 --> src/main.rs:5:18
  |
5 |         let _z = &mut *y;
  |                  ^^^^^^^
  |                  |
  |                  trying to reborrow alloc953+0x0 with Unique permission, but parent tag <1981> only grants SharedReadOnly permission for this memory
  |                  this is a reborrow of alloc953[0x0..0x4]
  |
```

I prefer this one, but it's hard to tell if I'm abusing the error reporting machinery. Though I'm not sure how much of a leg the existing code has to stand on, seeing that it pastes the whole error message into the title ;)

RalfJung replied:
No strong opinion on my side. However I think I'd prefer wording like "this is part of a reborrow at alloc953[0x0..0x4]" or so.

It also seems a bit odd to sometimes use `[...]` and sometimes `+` to join the index to the allocation ID?

saethlin replied:
Pronouns are hard. I wanted "this" to refer to the span, but I'm going to go with "this error occurs as part of a..." so there's no ambiguity.

You're right. I simply did not think of putting one value in the brackets.

RalfJung commented on src/stacked_borrows.rs:
```diff
@@ -621,6 +694,21 @@ trait EvalContextPrivExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         let this = self.eval_context_mut();
         if size == Size::ZERO {
             // Nothing to do for zero-sized accesses.
+            // But we still create a tag for them so that it's clear if a user misuses a pointer
```
This is confusing, since the code here does not create a tag.

Also I think the code duplication for `CreatedPointerTag` should be avoided.

RalfJung commented:
I started to review this, but then noticed I am confused since I thought the main motivation is about better errors when a tag is *missing*. But then the PR also does a lot of things with `CreatedPointerTag`, introducing code duplication and odd functions like `kind_to_perm_approximate` that I am not sure really make sense. I think I would prefer for these changes to be done separately; in particular I am not convinced the new tag creation message is really better (printing "approximate" information seems potentially more damaging than helpful).

So, could you reduce this PR to just the part about better diagnostics on missing tags, and make a separate PR for the tag creation message change, including a motivation for that change? In particular, as you noticed, there is no such thing as the single permission that is granted to a new tag -- the same tag can have *different* permissions for different parts of memory, and when that is the case, the message needs to reflect this to avoid being plain wrong.

> I'm also getting rid of the phrase "the borrow stack" in errors. From my own experience certainly and I think from my experience talking to people on the community Discord, it seems like people initially have the impression that there is one borrow stack. So a diagnostic that says "does not exist in the borrow stack" when tag-tracking only shows a creation of the tag and never removing it can be quite confusing.

I feel like the borrow stack is a core concept of Stacked Borrows (the name kind of gives it away ;), so I am not convinced that removing it from the messages is really helpful. What do you think about rewording the message to make it clear that this is "the borrow stack for location alloc1337+42", which hopefully should clarify that there is no single global borrow stack?

RalfJung commented:
> You of course know SB better than I do, but this looks to me like a bug in Miri, probably a bug in this PR. I feel like a pointer that grants Unique over alloc100517[0x0..0x2] should be able to be reborrows into SRO over that same memory range.

My guess would be that the parent tag simply does not exist in the borrow stack any more, and hence also cannot grant that permission.

saethlin commented:
> What do you think about rewording the message to make it clear that this is "the borrow stack for location alloc1337+42", which hopefully should clarify that there is no single global borrow stack?

I'm really struggling to do this. I made a point of wording the diagnostics as "Tried to do X but not possible because Y" because I think that is a very easy form to read, and for the most part trying to mention borrow stacks seems to make them convoluted. Can you suggest a diff?

RalfJung commented on src/stacked_borrows.rs:
```diff
@@ -338,16 +338,20 @@ impl<'tcx> Stack {
             err_sb_ub(format!(
-                "no item granting {} to tag {:?} at {:?} found in borrow stack.",
-                access, tag, dbg_ptr,
+                "no item granting {} to tag {:?} over {}[{:#x}..{:#x}] found in borrow stack.",
```
This still says "borrow stack" so I guess that alleviates my concern about mentioning that term? I am a bit confused now.^^

RalfJung replied:
Though this makes me wonder if this should also be reworded to be consistent with your new reborrowing errors?

saethlin replied:
I did my best to reword where I could in response to your comment about mentioning borrow stacks.

The code missing the borrow stack terminology is in `grant_error` now.

RalfJung replied:
I think I see what you mean.

I guess I came around to not having as strong an opinion about this as I used to. If you want to instead word things like "but parent tag {} is invalid for this memory range" (which seems more consistent with "parent tag {:?} grants insufficient permission over this memory range"), that is fine for me.

RalfJung commented on src/stacked_borrows.rs:
```diff
@@ -467,6 +469,41 @@ impl<'tcx> Stack {
+        );
+        if derived_from == SbTag::Untagged {
+            err_sb_ub(format!(
+                "{}, but parent tag {:?} does not grant the required permission over memory range",
```
Why does the untagged message not distinguish between the tag existing and not existing?

RalfJung commented:
> I'm really struggling to do this. I made a point of wording the diagnostics as "Tried to do X but not possible because Y" because I think that is a very easy form to read, and for the most part trying to mention borrow stacks seems to make them convoluted. Can you suggest a diff?

Looking at the source, the message actually uses the term "borrow stack" where I would expect it to show up. :)

However, are the messages in the OP still up to date? In particular the last one seems odd, since in the code it looks like "does not grant the required permission over" is only printed for untagged parents.

saethlin commented:
Just over a week ago I was trying to turn off GitHub email notifications and I accidentally turned them _all_ off for a few days. I knew something would slip through. Gah!

I've updated the diagnostic you mentioned in the OP (and I rebased onto `master`).

RalfJung commented on src/stacked_borrows.rs:
```diff
@@ -467,6 +469,51 @@ impl<'tcx> Stack {
+        });
+        if let Some(max_granted) = available_permissions.get(0) {
+            err_sb_ub(format!(
+                "{}, but parent tag {:?} grants insufficient permission: {:?}",
```
"insufficient permission: Unique" and "insufficient permission: SharedReadWrite" seems impossible, doesn't it?

saethlin replied:
Sure does.

RalfJung commented on src/stacked_borrows.rs:
```diff
@@ -467,6 +469,51 @@ impl<'tcx> Stack {
+        alloc_range: AllocRange,
+    ) -> InterpError<'static> {
+        let action = format!(
+            "trying to reborrow for {:?} over {:?}[{:#x}..{:#x}]",
```
"over" a memory range sounds like an odd choice of proposition to me? I am not a native speaker, but I would have expected "in" or "at".

saethlin replied:
Preposition choice is one of the most arcane parts of English. I don't think there are any choices I'm confident will read well to all users, so I'm trying to refactor it out.

RalfJung commented on src/stacked_borrows.rs:
```diff
@@ -338,16 +338,20 @@ impl<'tcx> Stack {
             err_sb_ub(format!(
-                "no item granting {} to tag {:?} at {:?} found in borrow stack.",
-                access, tag, dbg_ptr,
+                "no item granting {} to tag {:?} exists in the borrow stack for memory range {}[{:#x}..{:#x}].",
```
Potential alternative that seems more consistent with your other messages?
"trying {access} with tag {:?} at {:?}[{:#x}..{:#x}], but ,something something>"

In fact, even here it makes sense to distinguish "tag is in stack but is read-only", and "tag is not in stack", doesn't it?

RalfJung commented on src/diagnostics.rs:
```diff
@@ -261,6 +268,7 @@ fn report_msg<'tcx>(
     diag_level: DiagLevel,
     title: &str,
     span_msg: String,
+    labels: Vec<String>,
```
`span_msg` and `labels` are referring to the same thing, aren't they?

saethlin replied:
No, sadly. This is my hack to get the line break that I want into the span message, so that they look like this:
```
5 |         let _z = &mut *y;
  |                  ^^^^^^^
  |                  |
  |                  trying to reborrow alloc953+0x0 with Unique permission, but parent tag <1981> only grants SharedReadOnly permission for this memory
  |                  this is a reborrow of alloc953[0x0..0x4]
```
If I just add a newline into `span_msg`, I end up with this:
```
5 |         let _z = &mut *y;
  |                  ^^^^^^^
  |                  |
  |                  trying to reborrow alloc953+0x0 with Unique permission, but parent tag <1981> only grants SharedReadOnly permission for this memory
this is a reborrow of alloc953[0x0..0x4]
  |                  
```

RalfJung replied:
What I meant is, they could become a single `Vec<String>` -- `span_msg` is treated like the labels are?

I called it `span_msg` back then since the term "label" meant nothing to me -- it did not say where the label was put.

saethlin replied:
Oh! I see. Adjusted.

RalfJung commented on src/diagnostics.rs:
```diff
@@ -279,6 +287,12 @@ fn report_msg<'tcx>(
+    // Add extra labels to the primary span.
+    for label in labels {
+        err.span_label(span, label);
+    }
```
Further up we have some extra logic to make sure the `span_msg` is visible even on a `DUMMY_SP`... that should probably also be taken into account for the extra labels?

saethlin replied:
Yes, it probably should. Do you have any examples of cases where we can't get a span for the primary error? Things will probably look pretty chaotic anyway, but it would be nice to see.

RalfJung replied:
I'm afraid I don't know any examples. :/

saethlin replied:
Well, I haven't run into any so I'm going to declare this rare enough to not worry too much about for the moment.

RalfJung replied:
Yeah, as long as the existing logic is roughly preserved I think this is fine.

RalfJung commented on src/stacked_borrows.rs:
```diff
@@ -338,18 +345,15 @@ impl<'tcx> Stack {
         access: AccessKind,
         tag: SbTag,
-        dbg_ptr: Pointer<AllocId>, // just for debug printing amd error messages
+        (alloc_id, range, offset): (AllocId, AllocRange, Size), // just for debug printing amd error messages
```
```suggestion
        (alloc_id, range, offset): (AllocId, AllocRange, Size), // just for debug printing and error messages
```
(old typo that I probably wrote.^^)

Somewhere (in the doc comment above?) it should explain that the `range` is for the *entire* access/reborrow, and `offset` is the location of the stack that we currently act on.

RalfJung commented on src/stacked_borrows.rs:
```diff
@@ -467,6 +469,91 @@ impl<'tcx> Stack {
+    }
+
+    /// Report a descriptive error when `access` is not permitted based on `tag`.
+    fn access_error(
```
Looks like there is quite a bit of duplication between these two functions? Is it possible and useful to share some code here?

saethlin replied:
Took a shot at some deduplication. It sort of looks better?

RalfJung commented on src/stacked_borrows.rs:
```diff
@@ -467,6 +469,91 @@ impl<'tcx> Stack {
+            err_sb_ub(
+                format!(
+                    "{}, but {:?} only grants SharedReadOnly permission for this memory",
+                    action, tag
```
Seems a bit odd to repeat the tag (it also occurs at the end of `action`)

saethlin commented:
Along with addressing your feedback, I squashed this back down and rebased it onto `master`. I'm working on some other branches and I didn't want to deal with swapping between miri toolchains.

RalfJung commented on src/stacked_borrows.rs:
```diff
@@ -334,22 +341,21 @@ impl<'tcx> Stack {
     /// Test if a memory `access` using pointer tagged `tag` is granted.
     /// If yes, return the index of the item that granted it.
+    /// `range` refers the entire operation, and `offset` refers to the specific location in
+    /// `range` that we are currently checking.
```
`offset` is relative to `alloc_id` though, not relative to `range`, right?

saethlin replied:
Correct. My initial wording was ambiguous.

RalfJung commented on src/stacked_borrows.rs:
```diff
@@ -467,6 +473,72 @@ impl<'tcx> Stack {
+        error_offset: Size,
+    ) -> InterpError<'static> {
+        let action = format!(
+            "trying to reborrow {:?}[{:#x}] with {:?} permission via tag {:?}",
```
"via tag" sounds odd to me, I think something about "derived from tag" or "from parent tag" or so is more evocative.

saethlin replied:
I'm concerned that both "derived from tag" and "from parent tag" are both too suggestive that only the permission is from the tag.

I went back and forth with some other people and came up with other wordings which are very similar to your original ones :joy:

RalfJung commented on src/stacked_borrows.rs:
```diff
@@ -467,6 +473,72 @@ impl<'tcx> Stack {
+        if self.borrows.iter().any(|item| item.tag == tag && item.perm != Permission::Disabled) {
+            ", but that tag only grants SharedReadOnly permission for this memory"
+        } else {
+            ", but that tag does not exist in the borrow stack for this memory"
```
"this location" or "this memory location" sounds more natural than "this memory"? (At least I don't usually think of memory as something countable, grammatically it works more like "sand" for me.)

saethlin replied:
I'm going with "this location" in a desperate attempt to save on line length

RalfJung commented on src/diagnostics.rs:
```diff
@@ -279,18 +279,19 @@ fn report_msg<'tcx>(
         DiagLevel::Note => tcx.sess.diagnostic().span_note_diag(span, title),
     };
 
+    let first_line = span_msg.get(0).cloned().unwrap_or_default();
```
Why single out the first line?

You can just add the `err.note("(no span available)");` after all of them IMO, keeping the code simpler.

saethlin replied:
:facepalm:

RalfJung commented:
Yeah I think I like these wordings you have now. :)

saethlin commented:
Squashed down and apparently I got too fancy somewhere so this is now rebased onto `master`.

oli-obk commented:
@bors r=RalfJung

bors commented:
:pushpin: Commit 730cd272481427149d8d4a1e471d983a73f77105 has been approved by `RalfJung`

<!-- @bors r=RalfJung 730cd272481427149d8d4a1e471d983a73f77105 -->
<!-- homu: {"type":"Approved","sha":"730cd272481427149d8d4a1e471d983a73f77105","approver":"RalfJung"} -->

bors commented:
:hourglass: Testing commit 730cd272481427149d8d4a1e471d983a73f77105 with merge 670dc7d55126cad910e554d98ad6f49e5381f645...
<!-- homu: {"type":"BuildStarted","head_sha":"730cd272481427149d8d4a1e471d983a73f77105","merge_sha":"670dc7d55126cad910e554d98ad6f49e5381f645"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/miri/runs/5586868933?check_suite_focus=true)
Approved by: RalfJung
Pushing 670dc7d55126cad910e554d98ad6f49e5381f645 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"RalfJung","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/miri/runs/5586868933?check_suite_focus=true"},"merge_sha":"670dc7d55126cad910e554d98ad6f49e5381f645"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/miri/runs/5586868933?check_suite_focus=true)
Approved by: RalfJung
Pushing 670dc7d55126cad910e554d98ad6f49e5381f645 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"RalfJung","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/miri/runs/5586868933?check_suite_focus=true"},"merge_sha":"670dc7d55126cad910e554d98ad6f49e5381f645"} -->
